src/js/libgeo/GeoOps.js

```diff
@@ -3060,6 +3068,25 @@ geoOps.Poly.updatePosition = function(el) {
     }));
 };
 
+geoOps._helper.snapPointToLine = function(pos, line) {
+    // fail safe for far points
+    if (CSNumber._helper.isAlmostZero(pos.value[2])) return pos;
+    // project point to line - useful for semi free elements
+    var projPos = geoOps._helper.projectPointToLine(pos, line);
+    projPos = List.normalizeZ(projPos);
+
+    var sx = projPos.value[0].value.real;
+    var sy = projPos.value[1].value.real;
+    var rx = Math.round(sx / csgridsize) * csgridsize;
+    var ry = Math.round(sy / csgridsize) * csgridsize;
+    var newpos = List.realVector([rx, ry, 1]);
+    if (Math.abs(rx - sx) < 0.2 && Math.abs(ry - sy) < 0.2 &&
+        CSNumber.abs(List.scalproduct(line, newpos)).value.real < CSNumber.eps) {
+        pos = newpos;
```

Comment:
A feeling tells me that `pos = geoOps._helper.projectPointToLine(newpos, line);` might be better here for the stability of consecutive operations.